src/rdf4cpp/rdf/parser/private/IStreamQuadIteratorSerdImpl.cpp

```diff
@@ -0,0 +1,247 @@
+        std::ostringstream oss;
+        oss << prefix_it->second << suffix;
+        return IRI{oss.view(), this->node_storage};
+    } else {
+        return nonstd::make_unexpected(SerdStatus::SERD_ERR_BAD_CURIE);
+    }
+}
+
+Literal IStreamQuadIterator::Impl::get_literal(SerdNode const *literal, SerdNode const *datatype, SerdNode const *lang) const {
+    auto const literal_value = node_into_string_view(literal);
+
+    if (datatype != nullptr) {
+        return Literal{literal_value, IRI{node_into_string_view(datatype), this->node_storage}, this->node_storage};
+    } else if (lang != nullptr) {
+        return Literal{literal_value, node_into_string_view(lang), this->node_storage};
+    } else {
+        return Literal{literal_value, this->node_storage};
+    }
+}
+
+SerdStatus IStreamQuadIterator::Impl::on_error(void *voided_self, SerdError const *error) noexcept {
+    auto *self = reinterpret_cast<Impl *>(voided_self);
+
+    auto const buf_sz = vsnprintf(nullptr, 0, error->fmt, *error->args);
+    std::string message;
+
+    message.resize(buf_sz + 1);  // +1 for null-terminator
+    vsnprintf(message.data(), message.size(), error->fmt, *error->args);
+    message.resize(buf_sz - 1);  // drop null-terminator from vsnprintf and newline from serd
+
+    self->last_error = ParsingError{
+            .error_type = parsing_error_type_from_serd(error->status),
+            .line = error->line,
+            .col = error->col,
+            .message = message};
+
+    return SerdStatus::SERD_SUCCESS;
+}
+
+SerdStatus IStreamQuadIterator::Impl::on_base(void *voided_self, const SerdNode *uri) noexcept {
+    auto *self = reinterpret_cast<Impl *>(voided_self);
+    self->prefixes.emplace("", node_into_string_view(uri));
+    return SERD_SUCCESS;
+}
+
+SerdStatus IStreamQuadIterator::Impl::on_prefix(void *voided_self, SerdNode const *name, SerdNode const *uri) noexcept {
+    auto *self = reinterpret_cast<Impl *>(voided_self);
+    self->prefixes.emplace(node_into_string_view(name), node_into_string_view(uri));
+    return SERD_SUCCESS;
+}
+
+SerdStatus IStreamQuadIterator::Impl::on_stmt(void *voided_self,
+                                              SerdStatementFlags,
+                                              SerdNode const *graph,
+                                              SerdNode const *subj,
+                                              SerdNode const *pred,
+                                              SerdNode const *obj,
+                                              SerdNode const *obj_datatype,
+                                              SerdNode const *obj_lang) noexcept {
+
+    auto *self = reinterpret_cast<Impl *>(voided_self);
+
+    auto const graph_node = [&]() -> nonstd::expected<Node, SerdStatus> {
+        if (graph != nullptr) {
+            switch (graph->type) {
+                case SERD_CURIE:
+                    return self->get_prefixed_uri(pred);
```

Comment:
all `get_*` functions can throw if something malformed slipped through, e.g. "abc"^^xsd:int

That should be handled, probably here in `on_stmt`

Reply:
Ah right, totally forgot that

Reply:
But as far as I can tell only `get_literal` can throw right?
I mean the rest (IRI, BlankNode) is also not marked `noexcept` but as far as I can tell they don't validate anything

Reply:
Fixed it for literal and curies

Reply:
IRI and BlankNode don't validate yet but should at some point in the future.

Reply:
Having said that: we can take care of IRIs and BlankNodes when they got validation in place.

Reply:
I implemented it, because I didn't see your latest comment. I left TODO comments for IRI and BNodes